pkg/kubelet/allocation/allocation_manager.go

```diff
@@ -203,3 +287,213 @@ func (m *manager) SetActuatedResources(allocatedPod *v1.Pod, actuatedContainer *
+		// If there is a pending resize but the resize is not allowed, always use the allocated resources.
+		m.statusManager.SetPodResizePendingCondition(pod.UID, v1.PodReasonInfeasible, msg)
+		return podFromAllocation, nil
+	}
+
+	m.allocationMutex.Lock()
+	defer m.allocationMutex.Unlock()
+	// Desired resources != allocated resources. Can we update the allocation to the desired resources?
+	fit, reason, message := m.canResizePod(allocatedPods, pod, getNodeFunc)
+	if fit {
+		// Update pod resource allocation checkpoint
+		if err := m.SetAllocatedResources(pod); err != nil {
+			return nil, err
+		}
+		m.statusManager.ClearPodResizePendingCondition(pod.UID)
+
+		// Clear any errors that may have been surfaced from a previous resize. The condition will be
+		// added back as needed in the defer block, but this prevents old errors from being preserved.
+		m.statusManager.ClearPodResizeInProgressCondition(pod.UID)
+
+		return pod, nil
+	}
+
+	if reason != "" {
+		m.statusManager.SetPodResizePendingCondition(pod.UID, reason, message)
+	}
+
+	return podFromAllocation, nil
+}
+
+// canAdmitPod determines if a pod can be admitted, and gives a reason if it
+// cannot. "pod" is new pod, while "pods" are all admitted pods
+// The function returns a boolean value indicating whether the pod
+// can be admitted, a brief single-word reason and a message explaining why
+// the pod cannot be admitted.
+// allocatedPods should represent the pods that have already been admitted, along with their
+// admitted (allocated) resources.
+func (m *manager) canAdmitPod(allocatedPods []*v1.Pod, pod *v1.Pod) (bool, string, string) {
+	// Filter out the pod being evaluated.
+	allocatedPods = slices.DeleteFunc(allocatedPods, func(p *v1.Pod) bool { return p.UID == pod.UID })
+
+	// the kubelet will invoke each pod admit handler in sequence
+	// if any handler rejects, the pod is rejected.
+	// TODO: move out of disk check into a pod admitter
+	// TODO: out of resource eviction should have a pod admitter call-out
+	attrs := &lifecycle.PodAdmitAttributes{Pod: pod, OtherPods: allocatedPods}
+	for _, podAdmitHandler := range m.admitHandlers {
+		if result := podAdmitHandler.Admit(attrs); !result.Admit {
+			klog.InfoS("Pod admission denied", "podUID", attrs.Pod.UID, "pod", klog.KObj(attrs.Pod), "reason", result.Reason, "message", result.Message)
+
+			recordAdmissionRejection(result.Reason)
+			return false, result.Reason, result.Message
+		}
+	}
+
+	return true, "", ""
+}
+
+// canResizePod determines if the requested resize is currently feasible.
+// pod should hold the desired (pre-allocated) spec.
+// Returns true if the resize can proceed; returns a reason and message
+// otherwise.
+func (m *manager) canResizePod(allocatedPods []*v1.Pod, pod *v1.Pod, getNodeFunc func() (*v1.Node, error)) (bool, string, string) {
+	if v1qos.GetPodQOS(pod) == v1.PodQOSGuaranteed && !utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScalingExclusiveCPUs) {
```

Comment:
Not for this PR:

I think we can collapse canResizePod and canAdmitPod into a single function, by adding an operation enum into `lifecycle.PodAdmitAttributes`. The allocatable check is already a duplicate, so we'd just need a new admission filter for this check that only runs on resize. (we actually might want to scope some of the other checks to only run on add). Infeasible vs Deferred could be inferred from the admission reject reason.

```suggestion
	// TODO: Move this logic into a PodAdmitHandler by introducing an operation field to lifecycle.PodAdmitAttributes, and combine canResizePod with canAdmitPod.
	if v1qos.GetPodQOS(pod) == v1.PodQOSGuaranteed && !utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScalingExclusiveCPUs) {
```